include/aspect/simulator/assemblers/interface.h

```diff
@@ -0,0 +1,629 @@
+          Vector<double>              local_rhs;
+
+          /** Denotes which face matrices have actually been assembled in the DG field
+           * assembly. Entries for matrices not used (for example, those corresponding
+           * to non-existent subfaces; or faces being assembled by the neighboring cell)
+           * are set to false.
+           **/
+          std::vector<bool>               assembled_matrices;
+
+          /**
+           * Indices of those degrees of freedom that actually correspond
+           * to the temperature or compositional field. since this structure
+           * is used to represent just contributions to the advection
+           * systems, there will be no contributions to other parts of the
+           * system and consequently, we do not need to list here indices
+           * that correspond to velocity or pressure degrees (or, in fact
+           * any other variable outside the block we are currently considering)
+           */
+          std::vector<types::global_dof_index>   local_dof_indices;
+          /** Indices of the degrees of freedom corresponding to the temperature
+           * or composition field on all possible neighboring cells. This is used
+           * in the discontinuous Galerkin method. The outer std::vector has
+           * length GeometryInfo<dim>::max_children_per_face * GeometryInfo<dim>::faces_per_cell,
+           * and has size zero if in the continuous Galerkin case.
+           **/
+          std::vector<std::vector<types::global_dof_index> >   neighbor_dof_indices;
+        };
+      }
+    }
+  }
+
+  /**
+   * A namespace for the definition of assemblers for the various terms in the
+   * linear systems ASPECT solves.
+   */
+  namespace Assemblers
+  {
+    /**
+     * A base class for objects that implement assembly
+     * operations.
+     *
+     * The point of this class is primarily so that we can store
+     * pointers to such objects in a list. The objects are created
+     * in Simulator::set_assemblers() (which is the only place where
+     * we know their actual type) and destroyed in the destructor of
+     * the Simulator object.
+     */
+    template <int dim>
+    class Interface
+    {
+      public:
+        virtual ~Interface () {}
+
+        virtual
+        void
+        execute(internal::Assembly::Scratch::ScratchBase<dim> &,
+                internal::Assembly::CopyData::CopyDataBase<dim> &) const =0;
+
+        /**
+         * This function gets called if a MaterialModelOutputs is created
+         * and allows the assembler to attach AdditionalOutputs. The
+         * function might be called more than once for a
+         * MaterialModelOutput, so it is recommended to check if
+         * get_additional_output() returns an instance before adding a new
+         * one to the additional_outputs vector.
+         */
+        virtual void create_additional_material_model_outputs(MaterialModel::MaterialModelOutputs<dim> &) const {}
```

Comment:
I don't really understand how this works now. Before, this was a function that was part of the simulator, and now you moved it to the assembler interface, but only some assemblers implement it? For example, for Stokes only the StokesIncompressibleTerms. What if I don't use that assembler?
And you also changed some of the test results that tested if this is working correctly, so it would be great if you could explain this a bit more.

Reply:
If you do not use this assembler then this additional material model outputs object does not get created. But that is fine, each Assembler is only responsible for creating the additional material model outputs that it actually uses during assembly. This also means each assembler  has to check if the relevant additional material model outputs already exists (because another assembler already created them).